src/LinuxTracing/TracerImpl.cpp

```diff
@@ -294,35 +322,61 @@ bool TracerImpl::OpenUserSpaceProbes(const std::vector<int32_t>& cpus) {
     AddUretprobesFileDescriptors(uretprobes_fds_per_cpu, function);
     AddUprobesFileDescriptors(uprobes_fds_per_cpu, function);
 
-    for (const auto& [cpu, fd] : uretprobes_fds_per_cpu) {
-      uprobes_uretpobres_fds_per_cpu[cpu].push_back(fd);
-    }
-    for (const auto& [cpu, fd] : uprobes_fds_per_cpu) {
-      uprobes_uretpobres_fds_per_cpu[cpu].push_back(fd);
-    }
+    OpenRingBuffersOrRedirectOnExisting(uretprobes_fds_per_cpu, &fds_per_cpu_for_redirection,
+                                        &ring_buffers_, UPROBES_RING_BUFFER_SIZE_KB,
+                                        "uprobes_uretprobes");
+    OpenRingBuffersOrRedirectOnExisting(uprobes_fds_per_cpu, &fds_per_cpu_for_redirection,
+                                        &ring_buffers_, UPROBES_RING_BUFFER_SIZE_KB,
+                                        "uprobes_uretprobes");
   }
 
-  OpenUserSpaceProbesRingBuffers(uprobes_uretpobres_fds_per_cpu);
-
   return !uprobes_event_open_errors;
 }
 
-void TracerImpl::OpenUserSpaceProbesRingBuffers(
-    const absl::flat_hash_map<int32_t, std::vector<int>>& uprobes_uretpobres_fds_per_cpu) {
+bool TracerImpl::OpenUprobesToRecordStackOn(
+    const orbit_grpc_protos::FunctionToRecordStackOn& function, const std::vector<int32_t>& cpus,
+    absl::flat_hash_map<int32_t, int>* fds_per_cpu) {
   ORBIT_SCOPE_FUNCTION;
-  for (const auto& [/*int32_t*/ cpu, /*std::vector<int>*/ fds] : uprobes_uretpobres_fds_per_cpu) {
-    if (fds.empty()) continue;
+  const char* module = function.file_path().c_str();
+  const uint64_t offset = function.file_offset();
+  for (int32_t cpu : cpus) {
+    int fd;
+    fd = uprobes_with_stack_and_sp_event_open(module, offset, /*pid=*/-1, cpu, stack_dump_size_);
+    if (fd < 0) {
+      ORBIT_ERROR("Opening uprobe %s+%#x on cpu %d", function.file_path(), function.file_offset(),
+                  cpu);
+      return false;
+    }
+    (*fds_per_cpu)[cpu] = fd;
+  }
+  return true;
+}
 
-    // Create a single ring buffer per cpu.
-    int ring_buffer_fd = fds[0];
-    std::string buffer_name = absl::StrFormat("uprobes_uretprobes_%u", cpu);
-    ring_buffers_.emplace_back(ring_buffer_fd, UPROBES_RING_BUFFER_SIZE_KB, buffer_name);
+bool TracerImpl::OpenFunctionsToRecordStack(const std::vector<int32_t>& cpus) {
```

Comment:
OpenFunctionsToRecordStack**On**

Reply:
Actually I would call this `OpenUprobesToRecordStackOn` (we are not "opening a function").